### ISSUE TITLE
Rails' eager load hooks at not invoked

### ISSUE THREAD

**Issue by jonleighton** [mperham/sidekiq#1791]
Sidekiq [triggers eager loading by calling `Rails.application.eager_load!`](https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/cli.rb#L227). This causes all the files in `config.eager_load_paths` to be loaded, but it doesn't allow the `before_eager_load` hook to run, nor does it result in the `eager_load!` methods of the `config.eager_load_namespaces` being called.

To do that, we need `config.eager_load` to be set.

Rather than loading `config/environment.rb` only, Sidekiq could load `config/application.rb`, set `Rails.application.config.eager_load = true`, and then load `config/environment.rb`. I believe this would have the desired effect. [We do something similar in spring](https://github.com/rails/spring/blob/master/lib/spring/application.rb#L82-L92).

In the meantime, here's a workaround I'm using:

``` ruby
Sidekiq.configure_server do |config|
  Rails.application.config.eager_load = true
end
```

Cheers

**Comment by IAMRYO** [comment]
Using Rails 4.2.0 we still encountered this issue when explicitly requiring a file that also fell into the eager_load path.  Our solution was to add

```ruby
ActiveSupport::Dependencies.mechanism = :require
```
to our `sidekiq.rb` initializer, which will tell Rails autoload and eager_load to use `require` rather than `load`.

**Comment by jfelchner** [comment]
We are also getting a double-loading issue.  Tons of warnings about constants being reinitialized.

We have a requirement to use `require` in certain places in our code which is evidently causing this problem.  The app boots fine (no warnings) using only Rails but as soon as we `bundle exec sidekiq` it blows up.

We now have a requirement to `require` a couple of our models explicitly so that certain code is loaded before other code and now the app won't boot at all due to errors being raised instead of warnings.

**Comment by brianhempel** [comment]
We had a different issue caused by Sidekiq's hacky rails loading.

We set up a `config/initializers/sidekiq.rb` file with:

```ruby
Rails.application.configure do
  config.active_job.queue_adapter = :sidekiq
end
```

This initializer worked great for our web server and other processes. They enqueue jobs to Sidekiq fine. But we discovered that _within the Sidekiq worker process only_ the initializer was running but not taking effect. Within a `bundle exec sidekiq` process, the `ActiveJob::Base.queue_adapter` ended up as still the inline adapter, even though I can confirm that the above three lines of code were all being run during Sidekiq's startup.

Consequently, when a job tried to enqueue other jobs, they actually ran inline.

What fixed it for us was moving the code above to the `config/application.rb` file.

My understanding is that calling `Rails.application.configure` in an initializer is considered an acceptable practice, so it's unfortunate that in some cases Sidekiq seems to frustrate that pattern.

**Comment by ryansch** [comment]
I was also having issues with the way Sidekiq boots rails 4.  I ended up making the following change to my development.rb.

```ruby
config.cache_classes = !!Sidekiq.server?
config.eager_load = !!Sidekiq.server?
```

This had the benefit of running sidekiq in `allow_concurrency` mode.

It seems the existing sidekiq initializer runs too late for some of the boot code in our rails app.

**Comment by shime** [comment]
Had a custom directory inside `app/models` (eg. `app/models/foo-bar/`) that was throwing errors when starting Sidekiq in development.

```
wrong constant name Foo-bar
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/core_ext/module/qualified_const.rb:30:in `const_defined?'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/core_ext/module/qualified_const.rb:30:in `block in qualified_const_defined?'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/core_ext/module/qualified_const.rb:29:in `each'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/core_ext/module/qualified_const.rb:29:in `inject'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/core_ext/module/qualified_const.rb:29:in `qualified_const_defined?'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:376:in `qualified_const_defined?'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:145:in `block in watch_namespaces'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:143:in `map'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:143:in `watch_namespaces'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:643:in `new_constants_in'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:456:in `load_file'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:354:in `require_or_load'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:317:in `depend_on'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:233:in `require_dependency'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/engine.rb:472:in `block (2 levels) in eager_load!'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/engine.rb:471:in `each'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/engine.rb:471:in `block in eager_load!'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/engine.rb:469:in `each'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/engine.rb:469:in `eager_load!'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/engine.rb:346:in `eager_load!'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/application/finisher.rb:56:in `each'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/application/finisher.rb:56:in `block in <module:Finisher>'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/initializable.rb:30:in `instance_exec'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/initializable.rb:30:in `run'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/initializable.rb:55:in `block in run_initializers'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:226:in `block in tsort_each'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:348:in `block (2 levels) in each_strongly_connected_component'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:429:in `each_strongly_connected_component_from'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:347:in `block in each_strongly_connected_component'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:345:in `each'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:345:in `call'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:345:in `each_strongly_connected_component'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:224:in `tsort_each'
/Users/hrvoje/.rubies/ruby-2.2.0/lib/ruby/2.2.0/tsort.rb:203:in `tsort_each'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/initializable.rb:54:in `run_initializers'
/Users/hrvoje/.gem/ruby/2.2.0/gems/railties-4.2.0/lib/rails/application.rb:352:in `initialize!'
/Users/hrvoje/code/work/lessonhub/backend/config/environment.rb:5:in `<top (required)>'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:274:in `require'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:274:in `block in require'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:240:in `load_dependency'
/Users/hrvoje/.gem/ruby/2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:274:in `require'
/Users/hrvoje/.gem/ruby/2.2.0/gems/sidekiq-3.4.2/lib/sidekiq/cli.rb:241:in `boot_system'
/Users/hrvoje/.gem/ruby/2.2.0/gems/sidekiq-3.4.2/lib/sidekiq/cli.rb:50:in `run'
/Users/hrvoje/.gem/ruby/2.2.0/gems/sidekiq-3.4.2/bin/sidekiq:13:in `<top (required)>'
/Users/hrvoje/.gem/ruby/2.2.0/bin/sidekiq:23:in `load'
/Users/hrvoje/.gem/ruby/2.2.0/bin/sidekiq:23:in `<main>'
```

Doing exactly what @ryansch did resolved it.

**Comment by pmatsinopoulos** [comment]
I am late here. But can I ask why does `Sidekiq` asks Rails to work in eager load mode and does not let Rails environment decide about that? I am referring to this line here:

https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/cli.rb#L236

**Comment by mperham** [comment]
eager loading is required in Rails 4 because Sidekiq is multithreaded.

**Comment by pmatsinopoulos** [comment]
1) I guess that you are referring to the fact that Sidekiq might have many threads trying to load the same constant at the same time. In that case, having all the eager loaded paths constants loaded in advance would solve that problem.

IMHO, Sidekiq should allow Rails environment decide whether to eager load or not. For example, on `development` environment, not to eager load. Also, on `development` environment using 1 thread is pretty fine. But even if developer wants to use many threads, failing to load a constant by a thread due to contention by another, wouldn't happen too often, and wouldn't be of a big problem. Developer could live with that. On the other hand, on `production` environment, Rails eager loads anyway.

Please, note also that multithreaded problem in the context of loading of constants is not solved with the eager loading at Sidekiq (nor at Rails) level anyway. There still might be constants that are needed by Sidekiq, but Rails developer might not have put them inside the eager loaded paths.

In summary, I do not find a reason to make `cli` Sidekiq code ugly. Doesn't worth it. 

2) And there are warnings on `development` environment. With Sidekiq asking Rails to eager load, same file might be loaded multiple times, because `cache_classes` is `false`.  In case you want to keep the line that asks Rails to eager load, have you considered finding a way to set the `cache_classes` to `true` so that files are loaded only once (with `cache_classes` to `true`, they are `required` not `loaded`)?

**Comment by mperham** [comment]
@pmatsinopoulos I have no interest in changing the behavior for Rails 3 and 4.  You might be interested in this pull request https://github.com/mperham/sidekiq/pull/2457
